wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java

```diff
@@ -115,4 +142,149 @@ public void testOpenConnectionWithDataSourceClassName() throws SQLException {
+    }
+
+    assertThrows(SQLTransientConnectionException.class, () -> dataSource.getConnection());
+  }
+
+  /**
+   * After getting a successful connection from the pool, the connected instance becomes unavailable and the
+   * connection fails over to another instance through the Enhanced Failure Monitor.
+   */
+  @TestTemplate
+  @EnableOnDatabaseEngineDeployment(DatabaseEngineDeployment.AURORA)
+  @EnableOnTestFeature(TestEnvironmentFeatures.FAILOVER_SUPPORTED)
+  public void testEFMFailover() throws SQLException {
+    ProxyHelper.disableAllConnectivity();
+
+    List<TestInstanceInfo> instances = TestEnvironment.getCurrent().getInfo().getProxyDatabaseInfo().getInstances();
+
+    String writerIdentifier = instances.get(0).getInstanceId();
+    String readerIdentifier = instances.get(1).getInstanceId();
+    LOGGER.fine("Instance to connect to: " + writerIdentifier);
+    LOGGER.fine("Instance to fail over to: " + readerIdentifier);
+
+    ProxyHelper.enableConnectivity(writerIdentifier);
+    HikariDataSource dataSource = createDataSource(null);
+
+    // Get a valid connection, then make it fail over to a different instance
+    try (Connection conn = dataSource.getConnection()) {
+      assertTrue(conn.isValid(5));
+      String currentConnectionId = auroraUtil.queryInstanceId(conn);
+      assertTrue(currentConnectionId.equalsIgnoreCase(writerIdentifier));
+      LOGGER.fine("Connected to instance: " + currentConnectionId);
+
+      ProxyHelper.enableConnectivity(readerIdentifier);
+      ProxyHelper.disableConnectivity(writerIdentifier);
+
+      assertThrows(FailoverSuccessSQLException.class, () -> auroraUtil.queryInstanceId(conn));
+
+      // Check the connection is valid after connecting to a different instance
+      assertTrue(conn.isValid(5));
+      currentConnectionId = auroraUtil.queryInstanceId(conn);
+      LOGGER.fine("Connected to instance: " + currentConnectionId);
+      assertTrue(currentConnectionId.equalsIgnoreCase(readerIdentifier));
+
+      // Try to get a new connection to the failed instance, which times out
+      assertThrows(SQLTransientConnectionException.class, () -> dataSource.getConnection());
+    }
+  }
+
+  private HikariDataSource createDataSource(final Properties customProps) {
+    final HikariConfig config = getConfig(customProps);
+    HikariDataSource dataSource = new HikariDataSource(config);
+
+    final HikariPoolMXBean hikariPoolMXBean = dataSource.getHikariPoolMXBean();
+
+    LOGGER.fine("Starting idle connections: " + hikariPoolMXBean.getIdleConnections());
+    LOGGER.fine("Starting active connections: " + hikariPoolMXBean.getActiveConnections());
+    LOGGER.fine("Starting total connections: " + hikariPoolMXBean.getTotalConnections());
+    return dataSource;
+  }
+
+  private HikariConfig getConfig(Properties customProps) {
+    final HikariConfig config = new HikariConfig();
+    config.setUsername(TestEnvironment.getCurrent().getInfo().getProxyDatabaseInfo().getUsername());
```

Comment:
Can you store `TestEnvironment.getCurrent().getInfo().getProxyDatabaseInfo()` to a local variable to reduce code duplication?